### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by trevor-scheer**
Create:
getWeb3: Promise for getting web3 from the browser (adopted from Truffle)
Web3Provider:  Resolve, cache, and provide web3 and contracts to context
withWeb3: HOC for getting web3 from context and mapping the web3 object to props
withContract: HOC for adding a new contract to cache and mapping it to props

### CODE REVIEW

File: src/hocs/withContract.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import React, {Component} from "react";
+import PropTypes from "prop-types";
+
+const withContract = (contract, mapContractToProps) => WrappedComponent =>
```
Comment (evandavis) [inline src/hocs/withContract.js:4]: Alternate pattern that I think might be easier to reason about (and prevent you from needing to use `let`): 
1. Set the default of `mapContractToProps` to be a function that does the `else` case of your ternary below
2. Throw an error if `mapContractToProps` is anything but a function.

Comment (evandavis) [reply src/hocs/withContract.js:4]: Also, what's the expected type for `contract`?

File: src/hocs/withContract.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    componentDidMount() {
+      // If the contract doesn't exist in the cache then call back to have it resolved and cached.
+      if (!this.context.contracts[contract.contractName]) {
+        this.context.cacheNewContract(contract);
+      }
+    }
+
+    render() {
+      let contractProps = {};
+      if (this.context.contracts[contract.contractName]) {
```
Comment (evandavis) [inline src/hocs/withContract.js:20]: You do this lookup several times in this method, so it probably makes sense to put it in a variable (or a getter function.)

Let's say you store it as `contract`, it looks like your `resolved` prop would just be `!!contract`, which also prevents you from needing to clone the contract object.

File: src/hocs/withWeb3.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import React, {Component} from "react";
+import PropTypes from "prop-types";
+
+const withWeb3 = mapWeb3ToProps => WrappedComponent =>
```
Comment (evandavis) [inline src/hocs/withWeb3.js:4]: Same comment about using a default value for this argument and throwing an error if it's the wrong type.

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import _Web3Provider from "./provider/Web3Provider";
+import _withContract from "./hocs/withContract";
+import _withWeb3 from "./hocs/withWeb3";
+
+export const getWeb3 = _getWeb3;
+export const Web3Provider = _Web3Provider;
+export const withContract = _withContract;
+export const withWeb3 = _withWeb3;
+
+export default {
```
Comment (evandavis) [inline src/index.js:11]: Why the double export?

`export getWeb3 from "./utils/getWeb3";`

File: src/utils/getWeb3.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import Web3 from "web3";
+
+const getWeb3 = new Promise((resolve, reject) => {
```
Comment (evandavis) [inline src/utils/getWeb3.js:3]: Are you sure you want this to be a singleton? (Also, it doesn't look like `reject` is ever called)

**Review by evandavis — COMMENTED** [review]
Presented with the caveat that I don't understand what this thing actually does 🙃

File: src/hocs/withContract.js (hunk 1 above)
Comment (trevor-scheer) [reply src/hocs/withContract.js:4]: 1 + 2, 👍

3) `contract` is just an object with about 15 properties on it, always the same - should I just confirm `typeof object` or is there a way I can/should be more strict about that?

File: src/index.js (hunk 4 above)
Comment (trevor-scheer) [reply src/index.js:11]: Was going for the named + default exports but I can't really think of a good use case for the default.

File: src/utils/getWeb3.js (hunk 5 above)
Comment (trevor-scheer) [reply src/utils/getWeb3.js:3]: Good question - I hadn't considered this. This _does_ only ever need to resolved once, as it's just grabbing the web3 object off of the window or building its own if none is found. Is this a valid reason to leave it as is?

A higher order question in my head - are there reasons to keep a Promise like this as a singleton, or should I always be returning a new one? If so, what are the reasons to/not to?

**Pull request closed without merging**